### PR TITLE
Add RoundUpToMultiple and DivideAndRoundUp functions to MathUtils.h

### PR DESCRIPTION
**Pull request by amzn-tommy**
There are a number of minor, debatable things that came up which I don’t have a strong opinion on. I went with what felt best to me, but would appreciate feedback. In particular, I'd like people to be able to search for and find the functions if they are trying to do the same thing, without needing to roll their own due to lack of discoverability.

**Function names:**
considered RoundUpAndDivide, DivideByMultiple
IMO, DivideAndRoundUp does a better job of expressing what is going on.

considered RoundUp, RoundUpTo, RoundUpToMultiple
IMO, RoundUpToMultiple does a better job of expressing what you are rounding up to.

**Variable name for second parameter:**
considered roundUpTo, multiple, alignment
multiple would make it pretty clear that it is the ‘multiple’ in RoundUpToMultiple. But ‘alignment’ expresses how this function is used in many cases, to align memory allocations.

**Using a template to support multiple integer types:**
There is an Atom RHI version of DivideByMultiple that uses templates (I plan to deprecate this function following the process in https://github.com/o3de/sig-release/pull/23/files in favor of the AzCore version), but there is also precedence in MathUtils.h for implementing the function once for each type that is supported (e.g. for Lerp, GetAbs, etc. which have both float and double implementations). Since some types like floats would not be valid for this function, I went for copy-pasting implementations with the supported types.

Signed-off-by: Tommy Walton <waltont@amazon.com>

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h [hunk 1]
```diff
@@ -461,6 +461,36 @@ namespace AZ
...
+    constexpr uint32_t DivideAndRoundUp(uint32_t value, uint32_t alignment)
+    {
+        AZ_Assert(alignment != 0, "0 is an invalid multiple to round to.");
+        return ((value + alignment - 1) / alignment);
+    }
+
+    constexpr uint64_t DivideAndRoundUp(uint64_t value, uint64_t alignment)
+    {
+        AZ_Assert(alignment != 0, "0 is an invalid multiple to round to.");
+        return ((value + alignment - 1) / alignment);
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/Math/MathUtils.h:477]: uber nit: external parens are redundant (if this is a 'when in Rome' think and lots of other functions in this file do it it's fine to leave).

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h [hunk 2]
```diff
@@ -461,6 +461,36 @@ namespace AZ
...
     }
 
+    //! Returns the value divided by alignment, where the result is rounded up if the remainder is non-zero.
+    //! Example: alignment: 4
+    //! Value:  0 1 2 3 4 5 6 7 8
+    //! Result: 0 1 1 1 1 2 2 2 2
+    constexpr uint32_t DivideAndRoundUp(uint32_t value, uint32_t alignment)
+    {
+        AZ_Assert(alignment != 0, "0 is an invalid multiple to round to.");
+        return ((value + alignment - 1) / alignment);
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/Math/MathUtils.h:471]: question: Is it worth testing or asserting/warning about an input that is equal to INT_MAX (just thinking of potentially hairy edge cases people might inadvertently run into).

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h [hunk 3]
```diff
@@ -461,6 +461,36 @@ namespace AZ
...
+    }
+
+    constexpr uint64_t DivideAndRoundUp(uint64_t value, uint64_t alignment)
+    {
+        AZ_Assert(alignment != 0, "0 is an invalid multiple to round to.");
+        return ((value + alignment - 1) / alignment);
+    }
+    
+    //! Returns the value rounded up to a multiple of alignment
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/Math/MathUtils.h:480]: uber nit: Missing `.` at the end

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h [hunk 4]
```diff
@@ -461,6 +461,36 @@ namespace AZ
...
+    //! Example: roundTo: 4
+    //! Value:  0 1 2 3 4 5 6 7 8
+    //! Result: 0 4 4 4 4 8 8 8 8
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/Math/MathUtils.h:483]: These examples are great and super helpful to visualize what's happening

File: Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp [hunk 5]
```diff
@@ -58,4 +58,55 @@ namespace UnitTest
         EXPECT_NEAR(0.6, AZ::LerpInverse(1.0, 1.0 + 5.0 * epsilonD, 1.0 + 3.0 * epsilonD), epsilonD);
         EXPECT_NEAR(1.0, AZ::LerpInverse(1.0, 1.0 + 5.0 * epsilonD, 1.0 + 5.0 * epsilonD), epsilonD);
     }
+
+    class RoundUpToMultipleTestsFixture
```
Comment (hultonha) [inline Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp:62]: suggestion: Is it worth also having a sanity check on `DivideAndRoundUp` separately? It does not have to be as exhaustive and I realize it is being tests as `RoundUpToMultiple` is implemented in-terms of it, but a separate sanity check could be useful to ensure it gives the expected output

File: Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp [hunk 6]
```diff
@@ -58,4 +58,55 @@ namespace UnitTest
...
+            {
+                expectedOutput.push_back(i * alignment);
+            }
+        }
+
+        for (T inputValue = 0; inputValue < expectedOutput.size(); ++inputValue)
+        {
+            T result = RoundUpToMultiple(inputValue, alignment);
+            EXPECT_EQ(result, expectedOutput[inputValue]);
+        }
```
Comment (hultonha) [inline Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp:91]: suggestion: Rather than using a loop here it would be possible to use a matcher (e.g. `EXPECT_THAT(... ::testing::ElementsAreArray(expectedOutput)`) - the begin/end iterators could be used instead of the GTest doesn't play nicely with `AZStd::vector`

File: Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/Bits.h [hunk 7]
```diff
@@ -239,9 +240,9 @@ namespace AZ
         /**
         * Returns the value divided by alignment, where the result is rounded up if the remainder is non-zero.
         */
-        template <typename T> inline T DivideByMultiple(T value, size_t alignment)
+        template <typename T> inline T DivideByMultiple2(T value, size_t alignment)
```
Comment (hultonha) [inline Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/Bits.h:243]: question: I take it this is the call soon to be removed? Name could instead potentially be `DivideByMultiple_DEPRECATED` or `DivideByMultipleDeprecated` to be more explicit? Not a huge deal though and this isn't well defined/documented either so leaving as-is isn't a big deal

**Review by hultonha — APPROVED** [review]
This looks like a very useful addition to me! Awesome to take the time to extract this and make it easily available for others in future 🙂 Thanks!

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h (hunk 2 above)
Comment (santorac) [reply Code/Framework/AzCore/AzCore/Math/MathUtils.h:471]: > question: Is it worth testing or asserting/warning about an input that is equal to INT_MAX (just thinking of potentially hairy edge cases people might inadvertently run into).

Hm, more generally, you could run into issues if value + alignment > INT_MAX

File: Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp (hunk 5 above)
Comment (santorac) [reply Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp:62]: +1

File: Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp [hunk 8]
```diff
@@ -58,4 +58,55 @@ namespace UnitTest
...
+        uint64_t alignment = GetParam();
+        TestRoundUpToMultipleIsCorrect(alignment);
+    }
+    
+    INSTANTIATE_TEST_CASE_P(
+        MATH_RoundUpToMultiple,
+        RoundUpToMultipleTestsFixture,
+        // Multiples that we're going to test rounding up to
+        // Test with some low numbers, prime numbers, power of two numbers, and non-prime non-power-of-two numbers
+        ::testing::ValuesIn({1u,2u,3u,4u,5u,8u,9u,12u,13u}));
```
Comment (santorac) [inline Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp:111]: I understand that you are trying to test a wide range of data without lots of code, but this doesn't necessarily make for the best unit test. I think it's best practice to avoid logic and loops in unit tests. While it's good for _production code_ to be well-factored and general-purpose, doing that in unit tests makes code hard to read and could introduce bugs in your unit test that make it ineffective. Unit tests should be easy to read and unlikely to have bugs. Consider writing your unit tests like this instead:
```
EXPECT_EQ(RoundUpToMultiple(0, 1), 0);
EXPECT_EQ(RoundUpToMultiple(1, 1), 1);
EXPECT_EQ(RoundUpToMultiple(2, 1), 1);

EXPECT_EQ(RoundUpToMultiple(0, 2), 0);
EXPECT_EQ(RoundUpToMultiple(1, 2), 2);
EXPECT_EQ(RoundUpToMultiple(2, 2), 2);
EXPECT_EQ(RoundUpToMultiple(3, 2), 4);
EXPECT_EQ(RoundUpToMultiple(4, 2), 4);
EXPECT_EQ(RoundUpToMultiple(5, 2), 6);

EXPECT_EQ(RoundUpToMultiple(0, 13), 0);
EXPECT_EQ(RoundUpToMultiple(1, 13), 13);
EXPECT_EQ(RoundUpToMultiple(9, 13), 13);
EXPECT_EQ(RoundUpToMultiple(12, 13), 13);
EXPECT_EQ(RoundUpToMultiple(13, 13), 13);
EXPECT_EQ(RoundUpToMultiple(14, 13), 26);
EXPECT_EQ(RoundUpToMultiple(25, 13), 26);
EXPECT_EQ(RoundUpToMultiple(26, 13), 26);
EXPECT_EQ(RoundUpToMultiple(27, 13), 39);
```
Even though this doesn't cover as many values, IMO it's a better test because of its readability, low chance of unit test bugs, and even serves as documentation for the function.

Comment (santorac) [reply Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp:111]: It's also easier to validate this ^ though code review than the loops you have above

File: Gems/Atom/Asset/ImageProcessingAtom/Code/Source/Processing/PixelFormatInfo.cpp [hunk 9]
```diff
@@ -390,10 +390,11 @@ namespace ImageProcessingAtom
             return 0;
         }
 
-        // get number of blocks (ceiling round up for block count) and multiply with bits per block. Divided by 8 to get
+        // get number of blocks (divide image size by block size, then round up if there is a remainder to get block count) and multiply with bits per block. Divided by 8 to get
         // final byte size
```
Comment (santorac) [inline Gems/Atom/Asset/ImageProcessingAtom/Code/Source/Processing/PixelFormatInfo.cpp:394]: I think you can just delete this comment, it's redundant now.

File: Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/Bits.h (hunk 7 above)
Comment (santorac) [reply Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/Bits.h]: It seems there's a chance this could break since DivideByMultiple2 is templated and DivideAndRoundUp is not.

File: Gems/Atom/RHI/Code/Source/RHI.Reflect/ImageSubresource.cpp [hunk 10]
```diff
@@ -359,7 +359,7 @@ namespace AZ
             else
             {
                 const uint32_t bitsPerPixel = RHI::GetFormatSize(imageFormat) * 8;
-                subresourceLayout.m_bytesPerRow = (imageSize.m_width * bitsPerPixel + 7) / 8; // round up to nearest byte
+                subresourceLayout.m_bytesPerRow = DivideAndRoundUp(imageSize.m_width * bitsPerPixel, 8); // divide bit count by 8, then round up to nearest byte
```
Comment (santorac) [inline Gems/Atom/RHI/Code/Source/RHI.Reflect/ImageSubresource.cpp:362]: The comment is redundant now.

**Review by santorac — APPROVED** [review]
I left some suggestions but nothing that warrants blocking the PR.

File: Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/Bits.h (hunk 7 above)
Comment (hultonha) [reply Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/Bits.h]: Should the overloads handle this though? Thought maybe not every case for all sizes of `int`

**Review by AMZN-puvvadar — APPROVED** [review]
UdpTransport changes look good

File: Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/Bits.h [hunk 11]
```diff
@@ -241,7 +242,7 @@ namespace AZ
         */
         template <typename T> inline T DivideByMultiple(T value, size_t alignment)
```
Comment (pollend) [inline Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/Bits.h:243]: look like places where this is called in RHI has been replaced with AZ::DivideAndRoundUp. would it make sense to deprecate this?

File: Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp (hunk 8 above)
Comment (pollend) [reply Code/Framework/AzCore/Tests/Math/MathUtilsTests.cpp:111]: you can use a Parameterized test.

**Review by hultonha — APPROVED** [review]
Nice updates! 👍

**Review by pollend — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h [hunk 12]
```diff
@@ -461,6 +462,42 @@ namespace AZ
...
+            std::numeric_limits<uint32_t>::max() - value >= alignment,
+            "value '%" PRIu32 "' and alignment '%" PRIu32 "' will overflow when added together during DivideAndRoundUp.");
+        return (value + alignment - 1) / alignment;
+    }
+
+    constexpr uint64_t DivideAndRoundUp(uint64_t value, uint64_t alignment)
+    {
+        AZ_Assert(alignment != 0, "0 is an invalid multiple to round to.");
+        AZ_Assert(
+            std::numeric_limits<uint64_t>::max() - value >= alignment,
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Math/MathUtils.h:482]: AZStd::numeric_limits should be used.
It is the same as `std::numeric_limits` but brought into `AZStd` namespace.
It is available in AzCore/std/limits.h

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h [hunk 13]
```diff
@@ -461,6 +462,42 @@ namespace AZ
...
+            "value '%" PRIu64 "' and alignment '%" PRIu64 "' will overflow when added together during DivideAndRoundUp.");
+        return (value + alignment - 1) / alignment;
+    }
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Math/MathUtils.h:485]: These functions are in the same vein as SizeAlignUp/SizeAlignDown and PointerAlignUp/PointerAlignDown in [base.h](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/base.h#L180-L190)

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h (hunk 12 above)
Comment (amzn-tommy) [reply Code/Framework/AzCore/AzCore/Math/MathUtils.h]: I did not realize we had that. Will update.

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h [hunk 14]
```diff
@@ -468,16 +470,24 @@ namespace AZ
     constexpr uint32_t DivideAndRoundUp(uint32_t value, uint32_t alignment)
     {
         AZ_Assert(alignment != 0, "0 is an invalid multiple to round to.");
-        return ((value + alignment - 1) / alignment);
+        AZ_Assert(
+            AZStd::numeric_limits<uint32_t>::max() - value >= alignment,
+            "value '%" PRIu32 "' and alignment '%" PRIu32 "' will overflow when added together during DivideAndRoundUp.");
```
Comment (santorac) [inline Code/Framework/AzCore/AzCore/Math/MathUtils.h:475]: You forgot to include arguments for value and alignment

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h [hunk 15]
```diff
@@ -468,16 +470,24 @@ namespace AZ
...
+        return (value + alignment - 1) / alignment;
     }
 
     constexpr uint64_t DivideAndRoundUp(uint64_t value, uint64_t alignment)
     {
         AZ_Assert(alignment != 0, "0 is an invalid multiple to round to.");
-        return ((value + alignment - 1) / alignment);
+        AZ_Assert(
+            AZStd::numeric_limits<uint64_t>::max() - value >= alignment,
+            "value '%" PRIu64 "' and alignment '%" PRIu64 "' will overflow when added together during DivideAndRoundUp.");
```
Comment (santorac) [inline Code/Framework/AzCore/AzCore/Math/MathUtils.h:484]: Same here

**Review by santorac — CHANGES_REQUESTED** [review]

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h (hunk 14 above)
Comment (amzn-tommy) [reply Code/Framework/AzCore/AzCore/Math/MathUtils.h:475]: Great catch, thanks! Fixed.

File: Code/Framework/AzCore/AzCore/Math/MathUtils.h (hunk 15 above)
Comment (amzn-tommy) [reply Code/Framework/AzCore/AzCore/Math/MathUtils.h:484]: Fixed.

**Review by santorac — APPROVED** [review]
